File: SBiswas/Problems/Milestone1/Problem27.cpp
#include<iostream>
#include<vector>

using namespace std;


//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
    TreeNode* ans = nullptr;
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        bool done = false;

        inorder(original, cloned, target, done);

        return ans;
    }

    void inorder(TreeNode* root, TreeNode* clone_root, TreeNode* target, bool& done)
    {   
        if(done)
        {
            return;
        }

        if(!root)
        {
            return;
        }

        inorder(root->left, clone_root->left, target, done);
        if(root == target)
        {
            // clone_root is the answer
            done = true;
            ans = clone_root;
        }
        inorder(root->right, clone_root->right, target, done);
    }
};

File: Hritwik/Problems/Milestone1/Problem27.py
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        # Traverse on the cloned tree to get the target node and return that node
        queue=deque([cloned])
        while queue:
            for _ in range(len(queue)):
                node = queue.popleft()
                if target.val == node.val:
                    return node
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

# Create Original binary tree to traverse through it in an inorder traversal manner
#
#     Original    2
#              /     \
#             8        5  
#            / \        \
#           1   3        7

original = TreeNode(2)
original.left = TreeNode(8)
original.left.left = TreeNode(1)
original.left.right = TreeNode(3)
original.right = TreeNode(5)
original.right.right = TreeNode(7)
# Create a cloned tree to traverse through it in an inorder traversal manner
#
#        Cloned   2
#              /     \
#             8        5  
#            / \        \
#           1   3        7
cloned = TreeNode(2)
cloned.left = TreeNode(8)
cloned.left.left = TreeNode(1)
cloned.left.right = TreeNode(3)
cloned.right = TreeNode(5)
cloned.right.right = TreeNode(7)

solution = Solution()
target = TreeNode(8)
response = solution.getTargetCopy(original, cloned, target)
print(f'Value of the reference to the target node in the cloned tree.: {response.val}')
